### PR TITLE
Fixes #11547 Update exoplayer v2.15.0

### LINKED ISSUE #11547: Sound drops randomly when listening to audio messages
Hi. First, before getting to the annoying bug, I want to say thanks to the whole team for your great work :wave: :heart: 

### Bug description

When listening to audio messages, the sound very often drops randomly (every one to two minutes I would say). The app seems to still play the audio message, but no sound can be heard.
This has been the case since a few months. On my previous android phone (Wiko Lenny) too.
The problem does not happen on Signal for Linux Desktop.

### Steps to reproduce

Listening to some audio messages and wait for the problem to happen.

**Actual result:** The sound drops and I need to quit Signal and launch it again.
**Expected result:** The sound should still play until the message is over.

### Device info

**Device:** Oppo A54 5G. Model: CPH2195
**Android version:** Color OS V11.1, Android 11
**Signal version:** 5.19.4. Could also be reproduced on 5.20.4

### Link to debug log

https://debuglogs.org/de5b3b9ae24bf212a35e593131fef4000f56fb63acba217644e95dd7802139fc

### PR DESCRIPTION
**Pull request by makp9k**
Fixes #11547

### First time contributor checklist

- [x] I have read [how to contribute](https://github.com/signalapp/Signal-Android/blob/master/CONTRIBUTING.md) to this project
- [x] I have signed the [Contributor License Agreement](https://whispersystems.org/cla/)

### Contributor checklist

- [x] I am following the [Code Style Guidelines](https://github.com/signalapp/Signal-Android/wiki/Code-Style-Guidelines)
- [x] I have tested my contribution on these devices:
 * Sony Xperia XZ1, Android 9
 * OnePlus 7, Android 11
- [x] My contribution is fully baked and ready to be merged as is
- [x] I ensure that all the open issues my contribution fixes are mentioned in the commit message of my first commit using the `Fixes #1234` [syntax](https://help.github.com/articles/closing-issues-via-commit-messages/)

----------

### Description

This PR fixes an issue when VoiceNotePlaybackService loses its foreground state right after the playback starts, allowing the system to kill the service, which results in audio being interrupted. The changes are coming partly from the updated ExoPlayer (see https://github.com/google/ExoPlayer/issues/4988), so I decided to pick the latest version at the moment, which might become handy in further development.

### CODE REVIEW

File: app/src/main/java/org/thoughtcrime/securesms/components/voice/VoiceNoteMediaController.java [hunk 1]
```diff
@@ -312,7 +305,7 @@ private void cleanUpOldProximityWakeLockManager() {
   }
 
   private static boolean canExtractPlaybackInformationFromMetadata(@Nullable MediaMetadataCompat mediaMetadataCompat) {
-    return mediaMetadataCompat != null                  &&
+    return mediaMetadataCompat != null &&
```
Comment (alex-signal) [inline app/src/main/java/org/thoughtcrime/securesms/components/voice/VoiceNoteMediaController.java:308]: Alignment: In Signal Java, we align &&'s

File: app/src/main/java/org/thoughtcrime/securesms/video/VideoPlayer.java [hunk 2]
```diff
@@ -90,18 +82,21 @@ public VideoPlayer(Context context, AttributeSet attrs, int defStyleAttr) {
...
 
     if (exoPlayer == null) {
-      exoPlayer = ExoPlayerFactory.newSimpleInstance(context, renderersFactory, trackSelector, loadControl);
+      exoPlayer = new SimpleExoPlayer.Builder(context).build();
       exoPlayer.addListener(new ExoPlayerListener(this, window, playerStateCallback, playerPositionDiscontinuityCallback));
-      exoPlayer.addListener(new Player.EventListener() {
-        @Override
-        public void onPlayerStateChanged(boolean playWhenReady, int playbackState) {
+      exoPlayer.addListener(new Player.Listener() {
+        @Override public void onPlayWhenReadyChanged(boolean playWhenReady, int reason) {
```
Comment (alex-signal) [inline app/src/main/java/org/thoughtcrime/securesms/video/VideoPlayer.java:91]: For throughout the PR: Override should go on a separate line above signature.

File: app/src/main/java/org/thoughtcrime/securesms/video/VideoPlayer.java [hunk 3]
```diff
@@ -120,13 +115,12 @@ public void onPlayerStateChanged(boolean playWhenReady, int playbackState) {
 
     DefaultDataSourceFactory    defaultDataSourceFactory    = new DefaultDataSourceFactory(context, "GenericUserAgent", null);
     AttachmentDataSourceFactory attachmentDataSourceFactory = new AttachmentDataSourceFactory(context, defaultDataSourceFactory, null);
-    ExtractorsFactory           extractorsFactory           = new DefaultExtractorsFactory();
 
-    createMediaSource = () -> new ExtractorMediaSource.Factory(attachmentDataSourceFactory)
-                                                      .setExtractorsFactory(extractorsFactory)
-                                                      .createMediaSource(videoSource.getUri());
+    createMediaSource = () -> new ProgressiveMediaSource.Factory(attachmentDataSourceFactory)
+        .createMediaSource(MediaItem.fromUri(videoSource.getUri()));
```
Comment (alex-signal) [inline app/src/main/java/org/thoughtcrime/securesms/video/VideoPlayer.java:120]: Align `.`s

File: gradle/wrapper/gradle-wrapper.properties [hunk 4]
```diff
@@ -1,7 +1,6 @@
-#Wed Jul 21 15:41:52 ADT 2021
+#Sun Aug 22 21:27:30 CEST 2021
 distributionBase=GRADLE_USER_HOME
 distributionUrl=https\://services.gradle.org/distributions/gradle-6.7.1-all.zip
 distributionPath=wrapper/dists
 zipStorePath=wrapper/dists
-distributionSha256Sum=22449f5231796abd892c98b2a07c9ceebe4688d192cd2d6763f8e3bf8acbedeb
```
Comment (alex-signal) [inline gradle/wrapper/gradle-wrapper.properties:6]: We'll want to revert this file's changes.

**Review by alex-signal — CHANGES_REQUESTED** [review]
Wow! Thanks for doing this! Most of my code feedback is around code style and alignment.

We utilize ExoPlayer in a few different places, so it's important that all these paths are tested:

* Gif search
* Gif playback inside the conversation
* Gif playback after clicking a gif in the conversation
* Video playback throughout the app
* Video editing I *think*
* Sending a video (has some stuff around thumbnail generation, don't remember if this hits exo)

Also can you confirm that the playback speed and proximity sensor still functions as expected? (I see necessary changes to the dispatcher due to API changes)

File: gradle/wrapper/gradle-wrapper.properties (hunk 4 above)
Comment (makp9k) [reply gradle/wrapper/gradle-wrapper.properties:6]: My bad, I had no plans to include this in the PR :)

**Comment by makp9k** [comment]
Thank you for the review! I will check the points above and will update my PR accordingly.
Playback speed and proximity sensor were working on the devices I tested the app with.

File: app/src/main/java/org/thoughtcrime/securesms/components/voice/VoiceNoteMediaController.java (hunk 1 above)
Comment (makp9k) [reply app/src/main/java/org/thoughtcrime/securesms/components/voice/VoiceNoteMediaController.java:308]: This was already like this, didn't touch it, but will format properly.

File: app/src/main/java/org/thoughtcrime/securesms/conversation/ConversationItem.java [hunk 5]
```diff
@@ -146,6 +143,8 @@
 import java.util.Objects;
 import java.util.Set;
 
+import static org.thoughtcrime.securesms.util.ThemeUtil.isDarkTheme;
```
Comment (alex-signal) [inline app/src/main/java/org/thoughtcrime/securesms/conversation/ConversationItem.java:146]: We avoid static imports.

File: app/src/main/java/org/thoughtcrime/securesms/conversation/ConversationUpdateItem.java [hunk 6]
```diff
@@ -130,6 +126,7 @@ public void setEventListener(@Nullable EventListener listener) {
     this.eventListener = listener;
   }
 
+  @NotNull
   @Override
   public ConversationMessage getConversationMessage() {
```
Comment (alex-signal) [inline app/src/main/java/org/thoughtcrime/securesms/conversation/ConversationUpdateItem.java:131]: The pattern we use is:

```
<override annotation>
<access modifier> <nullability annotation> <return value> <method name>(<args>) {
```
```

File: app/src/main/java/org/thoughtcrime/securesms/conversation/ConversationUpdateItem.java [hunk 7]
```diff
@@ -22,12 +20,12 @@
 import com.google.android.material.button.MaterialButton;
 import com.google.common.collect.Sets;
 
+import org.jetbrains.annotations.NotNull;
```
Comment (alex-signal) [inline app/src/main/java/org/thoughtcrime/securesms/conversation/ConversationUpdateItem.java:23]: Wrong nullability annotation, as noted previously :)

File: app/src/main/java/org/thoughtcrime/securesms/giph/mp4/GiphyMp4ExoPlayerProvider.java [hunk 8]
```diff
@@ -6,39 +6,37 @@
...
  */
 final class GiphyMp4ExoPlayerProvider implements DefaultLifecycleObserver {
 
-  private final Context                  context;
-  private final TrackSelection.Factory   videoTrackSelectionFactory;
-  private final DefaultRenderersFactory  renderersFactory;
-  private final TrackSelector            trackSelector;
-  private final LoadControl              loadControl;
+  private final Context context;
+  private final OkHttpClient okHttpClient = ApplicationDependencies.getOkHttpClient().newBuilder().proxySelector(new ContentProxySelector()).build();
```
Comment (alex-signal) [inline app/src/main/java/org/thoughtcrime/securesms/giph/mp4/GiphyMp4ExoPlayerProvider.java:28]: alignment

File: app/src/main/java/org/thoughtcrime/securesms/giph/mp4/GiphyMp4ExoPlayerProvider.java [hunk 9]
```diff
@@ -6,39 +6,37 @@
...
+  private final DataSource.Factory dataSourceFactory = new ChunkedDataSourceFactory(okHttpClient, null);
+  private final MediaSourceFactory mediaSourceFactory = new ProgressiveMediaSource.Factory(dataSourceFactory);
 
   GiphyMp4ExoPlayerProvider(@NonNull Context context) {
-    this.context                    = context;
-    this.videoTrackSelectionFactory = new AdaptiveTrackSelection.Factory();
-    this.renderersFactory           = new DefaultRenderersFactory(context);
-    this.trackSelector              = new DefaultTrackSelector(videoTrackSelectionFactory);
-    this.loadControl                = new DefaultLoadControl();
+    this.context = context;
```
Comment (alex-signal) [inline app/src/main/java/org/thoughtcrime/securesms/giph/mp4/GiphyMp4ExoPlayerProvider.java:33]: alignment

File: app/src/main/java/org/thoughtcrime/securesms/giph/mp4/GiphyMp4ExoPlayerProvider.java [hunk 10]
```diff
@@ -6,39 +6,37 @@
...
   }
 
   @MainThread final @NonNull ExoPlayer create() {
-    SimpleExoPlayer exoPlayer = ExoPlayerFactory.newSimpleInstance(context, renderersFactory, trackSelector, loadControl);
+    SimpleExoPlayer exoPlayer = new SimpleExoPlayer.Builder(context)
```
Comment (alex-signal) [inline app/src/main/java/org/thoughtcrime/securesms/giph/mp4/GiphyMp4ExoPlayerProvider.java:37]: This `.` alignment is perfect :)

File: app/src/main/java/org/thoughtcrime/securesms/giph/mp4/GiphyMp4ProjectionPlayerHolder.java [hunk 11]
```diff
@@ -11,13 +11,12 @@
 import androidx.lifecycle.Lifecycle;
 
 import com.google.android.exoplayer2.ExoPlayer;
+import com.google.android.exoplayer2.MediaItem;
 import com.google.android.exoplayer2.Player;
-import com.google.android.exoplayer2.source.MediaSource;
 import com.google.android.exoplayer2.ui.AspectRatioFrameLayout;
 
-import org.signal.glide.Log;
+import org.jetbrains.annotations.NotNull;
```
Comment (alex-signal) [inline app/src/main/java/org/thoughtcrime/securesms/giph/mp4/GiphyMp4ProjectionPlayerHolder.java:18]: Wrong nullability annotation. If you want, there is a setting in AS that lets you choose a default nullability annotation.

File: app/build.gradle [hunk 12]
```diff
@@ -411,9 +411,10 @@ dependencies {
     implementation 'com.google.android.gms:play-services-maps:16.1.0'
     implementation 'com.google.android.gms:play-services-auth:16.0.1'
 
-    implementation 'com.google.android.exoplayer:exoplayer-core:2.9.1'
-    implementation 'com.google.android.exoplayer:exoplayer-ui:2.9.1'
-    implementation 'com.google.android.exoplayer:extension-mediasession:2.9.1'
+    implementation 'com.google.android.exoplayer:exoplayer-core:2.15.0'
+    implementation 'com.google.android.exoplayer:exoplayer-ui:2.15.0'
+    implementation 'com.google.android.exoplayer:extension-mediasession:2.15.0'
+    implementation 'androidx.media:media:1.4.1'
```
Comment (alex-signal) [inline app/build.gradle:417]: Is this a new requirement from exoplayer, or just for the Media Compat stuff? I'm surprised we didn't have this already.

**Review by alex-signal — CHANGES_REQUESTED** [review]
Hi! This is starting to look good. Mostly formatting changes and a few questions for my own edification. We do not use Jetbrains's nullability annotations, opting for Android's instead.

**Comment by makp9k** [comment]
@alex-signal please let me know if i can further improve it

**Comment by alex-signal** [comment]
Hi! Last ask, please run `./gradlew ktlintformat` :)

**Comment by makp9k** [comment]
Done :)

**Comment by alex-signal** [comment]
Thanks! I'll see to this getting merged in and let you know if we need anything else.

File: app/src/main/java/org/thoughtcrime/securesms/components/voice/VoiceNoteNotificationManager.java [hunk 13]
```diff
@@ -40,30 +35,26 @@
...
-      throw new IllegalArgumentException("Could not create a controller with given token");
-    }
+    controller = new MediaControllerCompat(context, token);
 
-    notificationManager = PlayerNotificationManager.createWithNotificationChannel(context,
-                                                                                  NotificationChannels.VOICE_NOTES,
-                                                                                  R.string.NotificationChannel_voice_notes,
-                                                                                  NOW_PLAYING_NOTIFICATION_ID,
-                                                                                  new DescriptionAdapter());
+    notificationManager = new PlayerNotificationManager.Builder(
```
Comment (alex-signal) [inline app/src/main/java/org/thoughtcrime/securesms/components/voice/VoiceNoteNotificationManager.java:44]: formatting

File: app/src/main/java/org/thoughtcrime/securesms/components/voice/VoiceNotePlaybackPreparer.java [hunk 14]
```diff
@@ -142,157 +130,147 @@ public void onTimelineChanged(Timeline timeline, @Nullable Object manifest, int
...
                        });
 
-                       player.prepare(dataSource);
+                       player.prepare();
                        canLoadMore = !singlePlayback;
                      } else if (Objects.equals(latestUri, uri)) {
                        Log.w(TAG, "Requested playback but no voice notes could be found.");
-                       ThreadUtil.postToMain(() -> Toast.makeText(context, R.string.VoiceNotePlaybackPreparer__failed_to_play_voice_message, Toast.LENGTH_SHORT)
-                                                        .show());
+                       ThreadUtil.postToMain(() -> Toast
```
Comment (alex-signal) [inline app/src/main/java/org/thoughtcrime/securesms/components/voice/VoiceNotePlaybackPreparer.java:137]: This is a situation where the alignment actually increases confusion. Should wrap this in braces and indent it.

File: app/src/main/java/org/thoughtcrime/securesms/components/voice/VoiceNotePlaybackPreparer.java [hunk 15]
```diff
@@ -142,157 +130,147 @@ public void onTimelineChanged(Timeline timeline, @Nullable Object manifest, int
...
-      throw new IllegalStateException("QueueDataAdapter and DataSource size inconsistency.");
-    }
   }
 
-  private @NonNull MediaDescriptionCompat createEndClone(@NonNull MediaDescriptionCompat source) {
-    return buildUpon(source).setMediaId("end").setMediaUri(END_URI).build();
+  private int indexOfPlayerMediaItemByUri(@NonNull Uri uri) {
+    for (int i = 0; i < player.getMediaItemCount(); i++) {
+      MediaItem.PlaybackProperties playbackProperties = player.getMediaItemAt(i).playbackProperties;
+      if (playbackProperties != null && playbackProperties.uri.equals(uri)) return i;
```
Comment (alex-signal) [inline app/src/main/java/org/thoughtcrime/securesms/components/voice/VoiceNotePlaybackPreparer.java:198]: Avoid same-line if where possible

**Review by alex-signal — CHANGES_REQUESTED** [review]
A few formatting changes, and there are also test failures. It's possible the tests failing are no longer relevant, and as such I'd ask they'd be removed. You can run tests on the command line via `./gradlew qa` or with normal command line gradle test commands for the `signal-android` module.

File: app/src/main/java/org/thoughtcrime/securesms/components/voice/VoiceNoteNotificationManager.java (hunk 13 above)
Comment (makp9k) [reply app/src/main/java/org/thoughtcrime/securesms/components/voice/VoiceNoteNotificationManager.java]: Fixed

File: app/src/main/java/org/thoughtcrime/securesms/components/voice/VoiceNotePlaybackPreparer.java (hunk 15 above)
Comment (makp9k) [reply app/src/main/java/org/thoughtcrime/securesms/components/voice/VoiceNotePlaybackPreparer.java]: Fixed

**Comment by makp9k** [comment]
Thank you for your time, Alex! I've adjusted the PR :)

**Comment by alex-signal** [comment]
Thank you!

This should go out in `5.23`

**Pull request closed without merging**